### DISCUSSION TITLE
How complex the `matchPackageNames` may be?

### DISCUSSION THREAD

**Discussion by ITaluone** [renovatebot/renovate#32750] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab

### Please tell us more about your question or problem

I have two package rules, One should group every package where package name contains `H.NotifyIcon` and one should group every package containing `Maui` but not `NotifyIcon`.

The interesting part of `renovate.json`:

```json    
    {
      "matchPackageNames": [
        "/H\\.NotifyIcon/"
      ],
      "groupName": "notify-icon"
    },
    {
      "matchPackageNames": [
        "/^(?:(?!NotifyIcon).)+(?:Maui)+.*$/"
      ],
      "groupName": "maui"
    }
```

As tested on https://regex101.com/ with this data
```
H.NotifyIcon
H.NotifyIcon.Maui
Microsoft.Maui.Controls
Microsoft.AspNetCore.Components.WebView.Maui
Microsoft.Maui.Controls.Compatibility
```
the regex is correct.

Why does this fail? See log output below..

### Logs (if relevant)

"error": {
         "validationSource": "renovate.json",
         "validationError": "The renovate configuration file contains some invalid settings",
         "validationMessage": "Invalid regExp for packageRules[1].matchPackageNames: `/^(?:(?!NotifyIcon).)+(?:Maui)+.*$/`",
         "message": "config-validation",
         "stack": "Error: config-validation\n    at mergeRenovateConfig (/builds/renovate-config/renovate-config/node_modules/renovate/lib/workers/repository/init/merge.ts:200:19)\n    at getRepoConfig (/builds/renovate-config/renovate-config/node_modules/renovate/lib/workers/repository/init/config.ts:14:12)\n    at initRepo (/builds/renovate-config/renovate-config/node_modules/renovate/lib/workers/repository/init/index.ts:[56](http://gitlab.aluone.local/renovate-config/renovate-config/-/jobs/9947#L56):12)\n    at Object.renovateRepository (/builds/renovate-config/renovate-config/node_modules/renovate/lib/workers/repository/index.ts:64:14)\n    at attributes.repository (/builds/renovate-config/renovate-config/node_modules/renovate/lib/workers/global/index.ts:202:11)\n    at start (/builds/renovate-config/renovate-config/node_modules/renovate/lib/workers/global/index.ts:187:7)\n    at /builds/renovate-config/renovate-config/node_modules/renovate/lib/renovate.ts:19:22"

**Comment by secustor** [comment]
We use the re2 library, which does not support look aheads.

**Reply by IT-VBFK** [reply]
Ok, any suggestions to come around this issue then?

**Reply by rarkins** [reply, marked as answer]
Use multiple array values in matchPackageNames, with one or more ! prefixes

**Reply by ITaluone** [reply]
Ahh, that's it. After a few tries I got it.

For anyone who is searching for something similar, here is what I ended up with:

```json
"packageRules": [
    {
      "matchPackageNames": [
        "/H.NotifyIcon/"
      ],
      "groupName": "H.NotifyIcon",
      "addLabels": ["H.NotifyIcon"]
    },
    {
      "matchPackageNames": [
        "!/NotifyIcon/",
        "/Maui/"
      ],
      "groupName": "Maui",
      "addLabels": ["MAUI"]
    }
  ]
```

Thanks for your help @secustor and @rarkins
